plugins/devlink/fu-devlink-netlink.c

```diff
@@ -0,0 +1,624 @@
+	if (tb[DEVLINK_ATTR_RELOAD_ACTION] != NULL)
+		g_strv_builder_add(
+		    tuples_builder,
+		    g_strdup_printf("reload_action=%u",
+				    mnl_attr_get_u8(tb[DEVLINK_ATTR_RELOAD_ACTION])));
+
+	return MNL_CB_OK;
+}
+
+static gchar *
+fu_devlink_netlink_get_event_id(FuDevlinkGenSocket *nlg, struct nlmsghdr *nlh)
+{
+	g_autoptr(GStrvBuilder) tuples_builder = g_strv_builder_new();
+	g_auto(GStrv) strv = NULL;
+
+	/* parse the message and build get event_id tuples */
+	mnl_cb_run(nlh,
+		   nlh->nlmsg_len,
+		   nlh->nlmsg_seq,
+		   nlh->nlmsg_pid,
+		   fu_devlink_netlink_event_id_msg_cb,
+		   tuples_builder);
+
+	strv = g_strv_builder_end(tuples_builder);
+
+	/* return the constructed event_id */
+	return g_strjoinv(",", strv);
+}
+
+/* receive and run callback on netlink messages */
+static gboolean
+fu_devlink_netlink_msg_recv_run(FuDevlinkGenSocket *nlg,
+				struct nlmsghdr *nlh,
+				mnl_cb_t cb,
+				void *data,
+				GError **error)
+{
+	FuDeviceEvent *event = NULL;
+	guint i = 0;
+	g_autofree gchar *event_id = NULL;
+	guint32 seq = nlh->nlmsg_seq;
+	gint rc;
+
+	/* generate event ID if device is available and we need emulation/recording */
+	if (nlg->is_emulated || nlg->save_events)
+		event_id = fu_devlink_netlink_get_event_id(nlg, nlh);
+
+	if (nlg->is_emulated) {
+		event = fu_device_load_event(nlg->device, event_id, error);
+		if (event == NULL)
+			return FALSE;
+	}
+
+	if (nlg->save_events)
+		event = fu_device_save_event(nlg->device, event_id);
+
+	do {
+		if (nlg->is_emulated) {
+			const guint8 *response_buf;
+			gsize response_len;
+			g_autofree gchar *response_key = g_strdup_printf("ResponseData%u", i++);
+			g_autoptr(GBytes) response_data =
+			    fu_device_event_get_bytes(event, response_key, NULL);
+
+			if (response_data == NULL)
+				return TRUE;
+
+			response_buf = g_bytes_get_data(response_data, &response_len);
+			if (response_len > (gsize)FU_DEVLINK_NETLINK_BUF_SIZE) {
+				g_set_error(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INVALID_DATA,
+					    "recorded response too large: %zu > %ld",
+					    response_len,
+					    FU_DEVLINK_NETLINK_BUF_SIZE);
+				return FALSE;
+			}
+			memcpy(nlg->buf, response_buf, response_len); /* nocheck:blocked */
+			rc = response_len;
+		} else {
+			rc = mnl_socket_recvfrom(nlg->nl, nlg->buf, MNL_SOCKET_BUFFER_SIZE);
+			if (rc == 0) {
+				return TRUE;
+			}
+			if (rc < 0) {
+				g_set_error(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_NOT_SUPPORTED,
+					    "failed to receive netlink message: %s",
+					    fwupd_strerror(errno));
+				return FALSE;
+			}
+			if (nlg->save_events) {
+				g_autoptr(GBytes) response_data = g_bytes_new(nlg->buf, rc);
+				g_autofree gchar *response_key =
+				    g_strdup_printf("ResponseData%u", i++);
+
+				fu_device_event_set_bytes(event, response_key, response_data);
+			}
+		}
+
+		rc = fu_devlink_netlink_msg_cb_run(nlg, rc, seq, cb, data, error);
+	} while (rc > MNL_CB_STOP);
+
+	return rc == MNL_CB_STOP;
```

Comment:
i think `return TRUE` would be more coverity friendly.

Reply:
done